feat(replay): add hooks for live status and polling of replay records

jerryzhou196 opened the pull request:
![output](https://github.com/user-attachments/assets/6171e914-3b53-4902-9ed6-b2de7e3b1626)

Background polls every 30 seconds and displays a `Refresh` button if it detects the `count_segments` of the rendered replay is different from the backend. 

Introduces two new hooks: 
- useIsLive to determine if a replay is currently live 
- usePollReplayRecord to poll the backend for updates on the replay record.

jerryzhou196 committed:
feat(replay): add hooks for live status and polling of replay records
Introduces two new hooks: useIsLive to determine if a replay is currently live and usePollReplayRecord to poll the backend for updates on the replay record.
The ReplayDetailsUserBadge component is updated to utilize these hooks, enhancing the live indicator functionality and allowing for automatic updates of replay data.

jerryzhou196 committed:
Merge branch 'master' into REPLAY-553-background-fetch-segment

jerryzhou196 committed:
fix(replays): update usePollReplayRecord to use useRef for isUpdated state

Refactored the usePollReplayRecord hook to utilize useRef for tracking the updated state of the replay record. This change ensures that the updated state persists across renders without causing unnecessary re-renders. Additionally, clarified the pollInterval comment for better understanding.

ryan953 commented on static/app/utils/replays/hooks/useIsLive.tsx:
```diff
@@ -0,0 +1,24 @@
+  replayReader: ReplayReader | null;
+};
+
+function useIsLive({replayReader}: Props) {
```
nit: you can put the export on the same line up here. Makes it easy to know that this is the main thing in the file right away.

```suggestion
export default function useIsLive({replayReader}: Props) {
```

ryan953 commented on static/app/utils/replays/hooks/useIsLive.tsx:
```diff
@@ -0,0 +1,24 @@
+  const [isLive, setIsLive] = useState<boolean>(replayReader?.getIsLive() ?? false);
+
+  if (isLive && replayReader) {
+    setTimeout(
+      () => {
+        setIsLive(false);
+      },
+      replayReader.getStartTimestampMs() + 60 * 60 * 1000 - Date.now()
+    );
```
The setTimeout should be inside a `useEffect` otherwise i'll be re-created each time this renders (or each time the parent renders)

Also, that useEffect should probably return a callback that calls`clearTimeout` when this gets unmounted.
Actually, we have a helper `useTimeout` that would help handle all the cases

jerryzhou196 committed:
refactor(replays): enhance useIsLive hook and update ReplayDetailsUserBadge component
- addressed Ryans timeout comment
- addressed Billys discussion with useQuery
- addressed Billys > 0 for timeout

billyvg commented on static/app/utils/replays/hooks/useIsLive.tsx:
```diff
@@ -0,0 +1,39 @@
+
+  if (replayReader) {
+    const TIME_UNTIL_REPLAY_FINISHED_MS =
+      replayReader.getStartTimestampMs() + 60 * 60 * 1000 - Date.now();
```
minimal savings but i would precalc the hours in ms, and  you can use `_` as a separator for readability.

```suggestion
      replayReader.getStartTimestampMs() + 3_600_000 - Date.now();
```

jerryzhou196 committed:
refactor(replays): improve time calculations and variable naming in hooks

- Updated time calculation in useIsLive to use a more readable format.
- Renamed replayRecord to polledReplayRecord in usePollReplayRecord for clarity.
- Removed unnecessary priority prop from RefreshButton in ReplayDetailsUserBadge component.

jerryzhou196 committed:
refactor(replays): update RefreshButton styling and add priority prop

- Added priority prop to the RefreshButton in ReplayDetailsUserBadge component for improved button emphasis.
- Removed custom background and hover color styles from RefreshButton to streamline its appearance.

jerryzhou196 committed:
Update RefreshButton styles and remove priority

Removed primary priority from RefreshButton and added custom styles.

jerryzhou196 committed:
added Jasmin's visual changes

jerryzhou196 committed:
Merge branch 'master' into REPLAY-553-background-fetch-segment

jerryzhou196 committed:
added Jesse's input
